### PR TITLE
Fix query_string_query to transform "foo:*" in an exists query on the field name

### PR DESCRIPTION
**Pull request by jimczi**
Currently "foo:\*" is parsed as prefix query on the field `foo` unless the field is defined in `default_field` or `fields`.
This commit fixes this behavior, "foo:\*" is now rewritten to an exists query on the field name.
This change also removes the assumption that "_all:\*" should return all docs.

relates #23356

### CODE REVIEW

**Comment by jpountz** [comment]
I like the idea! Does it also work if `_field_names` is disabled or should we only enable this if `_field_names` is enabled?

**Comment by jimczi** [comment]
> Does it also work if _field_names is disabled or should we only enable this if _field_names is enabled?

I pushed https://github.com/elastic/elasticsearch/pull/23433/commits/86d06973f3690332f71246eff95b429a6da69291 to handle this case

File: core/src/main/java/org/apache/lucene/queryparser/classic/ExistsFieldQueryExtension.java [hunk 1]
```diff
@@ -30,6 +32,13 @@
 
     @Override
     public Query query(QueryShardContext context, String queryText) {
-        return new ConstantScoreQuery(ExistsQueryBuilder.newFilter(context, queryText));
+        final FieldNamesFieldMapper.FieldNamesFieldType fieldNamesFieldType =
+            (FieldNamesFieldMapper.FieldNamesFieldType) context.getMapperService().fullName(FieldNamesFieldMapper.NAME);
+        if (fieldNamesFieldType.isEnabled() == false) {
+            // The field_names_field is disabled so we switch to a wildcard query that matches all terms
+            return new WildcardQuery(new Term(queryText, "*"));
```
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queryparser/classic/ExistsFieldQueryExtension.java]: that will only work on text/keyword fields? I think we need to perform this check up-front in the MappefFieldType?

Comment (jimczi) [inline core/src/main/java/org/apache/lucene/queryparser/classic/ExistsFieldQueryExtension.java]: True but that's also the current behavior. `foo:*` does not work if `foo` is not a text/keyword field. I don't get what you mean by check up-front ? Should we add another function in the MappedFieldType that returns match all query for the field ? Numbers would return open ranges query, text/keyword wildcard query and we would use this function only if the `_field_names` field is disabled ?

Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queryparser/classic/ExistsFieldQueryExtension.java]: Yeah sorry I thought we already had such a method, hence my comment! I think things are good this way, wildcard queries only make sense on text/keyword anyway. Thanks for the explanation!

**Review by jpountz — APPROVED** [review]
LGTM

**Comment by jimczi** [comment]
Thanks for reviewing @jpountz

**Pull request merged**